File: Data-Structures/Stack/stack.py
'''
Stack:

- A linear data structure which follows a particular order in which
the operations are performed.
- LIFO (Last In First Out)

Representation:

Access--> Node 2 ---------------------------- |-> data|next
          Node 1 ------------- |-> data|next -|     
          Node 0 -- data|next -|

Operations:
1. Push()
2. Pop()
3. GetPeek()
'''

class Node:
    def __init__(self, data= None):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.head = None
    
    #pushes the node at the top
    def push(self, node):
        node.next = self.head
        self.head = node
        print("Pushed!")

    #pop the node at the top
    def pop(self):
        if self.head is None:
            print("Empty list")
            return -1
        else:
            temp = self.head
            self.head = temp.next
            data = temp.data
            del temp
            return data
    
    #returns the data of the node at the top
    def getPeek(self):
        if self.head is None:
            print("Empty list")
            return -1
        else:
            data = self.head.data
            return data


s=Stack()
choice =0
while True:
    print("1.Push\t2.Pop\t3.Peak\t4.Stop ")
    choice = int(input("Enter choice "))
    node = Node()
    if choice==1:
        node.data = int(input("Enter data "))
        s.push(node)
    elif choice==2:
        result = s.pop()
        if result!=-1:
            print(result, " Popped")
    elif choice==3:
        result = s.getPeek()
        if result!=-1:
            print("Peek is ",result)
    elif choice==4:
        break
    else:
        print("Invalid choice")

File: Data-Structures/Stack/stack.cpp
#include<iostream>
using namespace std;

class Node{
    public:
    int data;
    Node * next;
    Node(){
        data = 0;
        next=NULL;
    }
};

class Stack{
    public:
    Node * head;

    Stack(){
        head=NULL;
    }
    Stack(Node *node){
        head=node;
    }

    //pushes the node at the top
    void push(Node *node){
        node->next = head;
        head = node;
        cout<<"Pushed!\n";
    }

    //pops the node at the top
    int pop(){
        if(head==NULL){
            cout<<"Empty stack";
            return -1;
        }
        else{
            Node * temp = head;
            head = temp->next;
            int popped = temp->data;
            free(temp);
            return popped;
        }
    }

    //returns the data of the node at the top
    int peek(){
        if(head==NULL){
            cout<<"Empty list\n";
            return -1;
        }
        else{
            return head->data;
        }
    }
};

int main(){

    Stack s;
    int data,choice;
    do{ 
        cout<<"\n1.Push\t2.Pop\t3.Peek\t4.Stop\n";
        cin>>choice;
        Node * node = new Node();
        switch(choice){
            case 1:
                cout<<"Enter data ";
                cin>>node->data;
                s.push(node);
                break;
            case 2:
                int result = s.pop();
                if(result !=-1){
                    cout<<result<<" Popped!";
                }
                break;
            case 3:
                int result = s.peek();
                if(result!=-1){
                    cout<<"Peek element is "<<s.peek();
                }
                break;
            case 4:
                break;
        }

    }while(choice!=4);
    return 0;
}